### PR TITLE
Follow speed limits for slotcar robots

### PR DESCRIPTION
**Pull request by luca-della-vedova**
### Implemented feature

This PR allows simulated robots to respect speed limits specified in path requests, requires https://github.com/open-rmf/rmf_internal_msgs/pull/28 and https://github.com/open-rmf/rmf_ros2/pull/132

### Implementation description

A new `std::optional<double> max_speed` is added to `UpdateResult` in `slotcar_common`, if a speed limit was specified in the message (specifically if it was > 0.0) the minimum between vehicle nominal speed and speed limit will be the navigation speed of the vehicle, before it was fixed at being the maximum robot speed at all times.

Since we need to store more information than a simple pose for differential drive robots' trajectories (i.e. speed limit), a new struct `DiffDriveTrajectory` was added and a minor refactoring took place to use it throughout the code.

### CODE REVIEW

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 1]
```diff
@@ -276,6 +279,8 @@ void SlotcarCommon::handle_ackermann_path_request(
   AckermannTrajectory traj(
     Eigen::Vector2d(locations[0].x, locations[0].y),
     Eigen::Vector2d(locations[1].x, locations[1].y));
+  if (locations[1].obey_approach_speed_limit)
+    traj.approach_speed_limit = locations[1].approach_speed_limit;
```
Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:283]: We should also check if `approach_speed_limit` is larger than 0.

Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:283]: `approach_speed_limit` should be check that it is larger than 0, as it looks like it is not checked further down the code path, https://github.com/open-rmf/rmf_simulation/blob/feature/speed_limits/rmf_robot_sim_common/src/utils.cpp#L32

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 2]
```diff
@@ -341,6 +346,8 @@ void SlotcarCommon::handle_ackermann_path_request(
         Eigen::Vector2d(wp[2].x(), wp[2].y()));
 
       AckermannTrajectory& last_traj = this->ackermann_trajectory.back();
+      if (locations[i].obey_approach_speed_limit)
+        sp2.approach_speed_limit = locations[i].approach_speed_limit;
```
Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:350]: `approach_speed_limit` should be check that it is larger than 0, as it looks like it is not checked further down the code path, https://github.com/open-rmf/rmf_simulation/blob/feature/speed_limits/rmf_robot_sim_common/src/utils.cpp#L32

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 3]
```diff
@@ -362,6 +369,9 @@ void SlotcarCommon::handle_ackermann_path_request(
         Eigen::Vector2d(tangent1.x(), tangent1.y()),
         Eigen::Vector2d(0, 0),
         true);
+      if (locations[i-1].obey_approach_speed_limit)
+        turn_traj.approach_speed_limit = locations[i-1].approach_speed_limit;
```
Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:373]: `approach_speed_limit` should be check that it is larger than 0, as it looks like it is not checked further down the code path, https://github.com/open-rmf/rmf_simulation/blob/feature/speed_limits/rmf_robot_sim_common/src/utils.cpp#L32

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 4]
```diff
@@ -370,6 +380,8 @@ void SlotcarCommon::handle_ackermann_path_request(
         Eigen::Vector2d(wp[2].x(), wp[2].y()));
       end_traj.v0 = wp1_to_wp2_norm;
       end_traj.v1 = wp1_to_wp2_norm;
+      if (locations[i].obey_approach_speed_limit)
+        end_traj.approach_speed_limit = locations[i].approach_speed_limit;
```
Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:384]: `approach_speed_limit` should be check that it is larger than 0, as it looks like it is not checked further down the code path, https://github.com/open-rmf/rmf_simulation/blob/feature/speed_limits/rmf_robot_sim_common/src/utils.cpp#L32

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 5]
```diff
@@ -223,8 +223,10 @@ void SlotcarCommon::handle_diff_drive_path_request(
 
     Eigen::Quaterniond quat(
       Eigen::AngleAxisd(msg->path[i].yaw, Eigen::Vector3d::UnitZ()));
-    trajectory.at(i).translation() = v3;
-    trajectory.at(i).linear() = Eigen::Matrix3d(quat);
+    trajectory.at(i).pose.translation() = v3;
+    trajectory.at(i).pose.linear() = Eigen::Matrix3d(quat);
+    if (msg->path[i].obey_approach_speed_limit)
+      trajectory.at(i).approach_speed_limit = msg->path[i].approach_speed_limit;
```
Comment (aaronchongth) [inline rmf_robot_sim_common/src/slotcar_common.cpp:229]: `approach_speed_limit` should be check that it is larger than 0, as it looks like it is not checked further down the code path, https://github.com/open-rmf/rmf_simulation/blob/feature/speed_limits/rmf_robot_sim_common/src/utils.cpp#L32

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Thanks for adding this feature, @luca-della-vedova! I've taken a quick look, other than the comments, I also feel that we should revisit functions like `calculate_control_signals` and `calculate_joint_control_signals` in the future and remove the use of default arguments in the next major versions release, as those arguments are used wholly by our downstream packages so far.

**Comment by luca-della-vedova** [comment]
Added checks for speed limit being greater than 0 https://github.com/open-rmf/rmf_simulation/pull/56/commits/02dc918ed81380f28c2b5b2269ad5f6b8c456771.

> I also feel that we should revisit functions like `calculate_control_signals` and `calculate_joint_control_signals` in the future and remove the use of default arguments in the next major versions release, as those arguments are used wholly by our downstream packages so far.

This can be done, so the idea is that all the arguments must be specified at all times instead of relying on default arguments? In that case I think some refactoring is also in order to reduce the number of input arguments to these functions, maybe by packing more motion parameters in `struct`s

**Comment by luca-della-vedova** [comment]
Closed via #71

**Pull request closed without merging**